protocol/icmpv6.go

```diff
@@ -0,0 +1,628 @@
+//    |                                                               |
+//    *                                                               *
+//    |                                                               |
+//    *                       Source Address [N]                      *
+//    |                                                               |
+//    *                                                               *
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |                                                               |
+//    .                                                               .
+//    .                         Auxiliary Data                        .
+//    .                                                               .
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type MLDv2Record struct {
+	Type             uint8
+	AuxDataLen       uint8 // this may be 0 to indicate the absence of any auxiliary data.
+	NumberOfSources  uint16
+	MulticastAddress net.IP
+	SourceAddresses  []net.IP
+	AuxData          []uint32 // NOT USED
+}
+
+func (r *MLDv2Record) Len() uint16 {
+	return 20 + uint16(r.AuxDataLen)*4 + r.NumberOfSources*16
+}
+
+func (r *MLDv2Record) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(r.Len()))
+	n := 0
+	data[n] = r.Type
+	n += 1
+	data[n] = r.AuxDataLen
+	n += 1
+	binary.BigEndian.PutUint16(data[n:], r.NumberOfSources)
+	n += 2
+	copy(data[n:n+4], r.MulticastAddress.To16())
+	n += 16
+	for _, src := range r.SourceAddresses {
+		copy(data[n:n+4], src.To16())
+		n += 16
+	}
+	for _, d := range r.AuxData {
+		binary.BigEndian.PutUint32(data[n:], d)
+		n += 4
+	}
+	return
+}
+
+func (r *MLDv2Record) UnmarshalBinary(data []byte) error {
+	if len(data) < 8 {
+		return errors.New("The []byte is too short to unmarshal a full MLDv2Record message.")
+	}
+	n := 0
+	r.Type = data[n]
+	n += 1
+	r.AuxDataLen = data[n]
+	n += 1
+	r.NumberOfSources = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	r.MulticastAddress = data[n : n+16]
+	n += 16
+	if len(data) < int(r.Len()) {
+		return fmt.Errorf("The []byte is too short to unmarshal a full MLDv2Record message.")
+	}
+	for i := uint16(0); i < r.NumberOfSources; i++ {
+		r.SourceAddresses = append(r.SourceAddresses, data[n:n+16])
+		n += 16
+	}
+	for i := uint8(0); i < r.AuxDataLen; i++ {
+		r.AuxData = append(r.AuxData, binary.BigEndian.Uint32(data[n:]))
+		n += 4
+	}
+	return nil
+}
+
+func NewMLDv2Record(recordType uint8, group net.IP, sources []net.IP) *MLDv2Record {
+	return &MLDv2Record{
+		Type:             recordType,
+		MulticastAddress: group,
+		NumberOfSources:  uint16(len(sources)),
+		SourceAddresses:  sources,
+	}
+}
+
+func NewICMPv6ByHeader(header *ICMPv6Header) util.Message {
```

Comment:
Given that we don't copy the header to the new message, I find the function a bit confusing. I suggest that you change the parameter to be the `Type` directly.

Reply:
I would update to read ICMPv6 type from data[n], and use it as the parameter directlly.